### PR TITLE
gateway-api: fix a race condition when updating status

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Fixes https://github.com/istio/istio/issues/49120

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
I think we can use one single atomic `statusEnabled` to protect

File: pilot/pkg/config/kube/gateway/controller.go [hunk 1]
```diff
@@ -247,14 +249,19 @@ func (c *Controller) QueueStatusUpdates(r GatewayResources) {
...
-			c.statusController.EnqueueStatusUpdateResource(ws.Unwrap(), res)
+			statusController := c.statusController.Load()
+			if statusController == nil {
+				// leader election loss
+				break
+			}
+			statusController.EnqueueStatusUpdateResource(ws.Unwrap(), res)
 		}
 	}
 }
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/controller.go:267]:
```suggestion
func (c *Controller) handleStatusUpdates(configs []config.Config) {
	statusController := c.statusController.Load()
	if statusController == nil || !c.statusEnabled.Load() {
		return
	}
	for _, cfg := range configs {
		ws := cfg.Status.(*kstatus.WrappedStatus)
		if ws.Dirty {
			res := status.ResourceFromModelConfig(cfg)
			statusController.EnqueueStatusUpdateResource(ws.Unwrap(), res)
		}
	}
}
``

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/controller.go:267]: Your code works as well I just don't think there is a need to do the checks multiple times like we had?

File: pilot/pkg/config/kube/gateway/controller.go [hunk 2]
```diff
@@ -247,14 +249,19 @@ func (c *Controller) QueueStatusUpdates(r GatewayResources) {
...
 		return
 	}
 	for _, cfg := range configs {
 		ws := cfg.Status.(*kstatus.WrappedStatus)
 		if ws.Dirty {
 			res := status.ResourceFromModelConfig(cfg)
-			c.statusController.EnqueueStatusUpdateResource(ws.Unwrap(), res)
+			statusController := c.statusController.Load()
+			if statusController == nil {
+				// leader election loss
```
Comment (jaellio) [inline pilot/pkg/config/kube/gateway/controller.go:261]: What are your thoughts on adding a debug or trace log here?

**Comment by dddddai** [comment]
> I think we can use one single atomic `statusEnabled` to protect

I think there are 2 approaches:
1. Use `c.statusEnabled` and just don't set `c.statusController` to nil when it loses leader election
2. Make R/W of `c.statusController` atomic. Then `c.statusEnabled` can be removed

**Review by howardjohn — APPROVED** [review]
thanks!

**Pull request merged**